src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs

```diff
@@ -19,63 +19,102 @@
+            IDictionary<Address, UserOperationPool> userOperationPools,
+            IDictionary<Address, UserOperationSimulator> userOperationSimulators,
             ISpecProvider specProvider,
+            IStateProvider stateProvider,
+            ISigner signer,
             ILogger logger)
         {
-            _userOperationTxBuilder = userOperationTxBuilder;
-            _userOperationPool = userOperationPool;
-            _userOperationSimulator = userOperationSimulator;
+            _userOperationTxBuilders = userOperationTxBuilders;
+            _userOperationPools = userOperationPools;
+            _userOperationSimulators = userOperationSimulators;
             _specProvider = specProvider;
+            _stateProvider = stateProvider;
+            _signer = signer;
             _logger = logger;
         }
 
         public IEnumerable<Transaction> GetTransactions(BlockHeader parent, long gasLimit)
         {
             IDictionary<Address, HashSet<UInt256>> usedAccessList = new Dictionary<Address, HashSet<UInt256>>();
-            IList<UserOperation> userOperationsToInclude = new List<UserOperation>();
+            // IList<UserOperation> userOperationsToInclude = new List<UserOperation>();
+            IDictionary<Address, IList<UserOperation>> userOperationsToIncludeByEntryPoint =
+                new Dictionary<Address, IList<UserOperation>>();
             ulong gasUsed = 0;
 
-            IEnumerable<UserOperation> userOperations =
-                _userOperationPool
+            IList<Tuple<Address, UserOperation>> combinedUserOperations = new List<Tuple<Address, UserOperation>>();
+            foreach (Address entryPoint in _userOperationPools.Keys)
+            {
+                IEnumerable<UserOperation> entryPointUserOperations = 
+                    _userOperationPools[entryPoint]
                     .GetUserOperations()
-                    .Where(op => op.MaxFeePerGas >= parent.BaseFeePerGas)
-                    .OrderByDescending(op => CalculateUserOperationPremiumGasPrice(op, parent.BaseFeePerGas));
-            foreach (UserOperation userOperation in userOperations)
+                    .Where(op => op.MaxFeePerGas >= parent.BaseFeePerGas);
+
+                foreach(UserOperation userOperation in entryPointUserOperations)
+                {
+                    combinedUserOperations.Add(Tuple.Create(entryPoint, userOperation));
+                }
+            }
+            IList<Tuple<Address, UserOperation>> sortedUserOperations = 
+                combinedUserOperations.OrderByDescending(
+                op => 
+                    CalculateUserOperationPremiumGasPrice(op.Item2, parent.BaseFeePerGas))
+                .ToList();
+            
+            foreach (Tuple<Address, UserOperation> addressedUserOperation in sortedUserOperations)
             {
-                if (gasUsed >= (ulong)gasLimit) continue;
+                UserOperation userOperation = addressedUserOperation.Item2;
```

Comment:
style: in recent csharp versions you can just do

```csharp
(UserOperation userOperation, Address entryPoint) = addressedUserOperation;
```

or just

```csharp
var (userOperation, entryPoint) = addressedUserOperation;
```

Both look more elegant IMHO.